### ISSUE TITLE
Finishing ... Done! with no plugins, any log file?

### ISSUE THREAD

**Issue by andreicristianpetcu** [junegunn/vim-plug#177]
Hi, 

I have a Centos 7 Vagrant box on which I run the next series of commands. It is based on this baseimage "https://f0fff3908f081cb6461b407be80daf97f07ac418.googledrive.com/host/0BwtuV7VyVTSkUG1PM3pCeDJ4dVE/centos7.box"
 
```bash
$sudo yum install -y vim-enhanced git
$python --version
Python 2.6.6
$ruby --version
ruby 1.8.7 (2011-06-30 patchlevel 352) [x86_64-linux]
$rm -rf ~/.vim && wget https://raw.githubusercontent.com/andreicristianpetcu/dotfiles/c6ed5186f8d0e1d0e31cded252185d40bb4fdb94/vimrc -O ~/.vimrc && vim
```
Vim fires up with the "PlugInstall" command as it should but I get this
```bash
Updated. Elapsed time: 0.036210 sec.
- Finishing ... Done!
```
```bash
$ls -l ~/.vim/bundle
total 0
```
```bash
:PlugStatus
Finished. 101 error(s). 
x fzf: 
   Not found. Try PlugInstall.
x unite-outline:
   Not found. Try PlugInstall.
```
How can I troubleshoot this? Is there any dependency that vim-plug requires except for git, ruby and python? This works on Ubuntu 14.04.

Thank you for building this awesome plugin!

**Comment by junegunn** [comment]
Are you running the latest version of vim-plug? This seems like a duplicate of #170 and #173.

**Comment by andreicristianpetcu** [comment]
I'm using this commint (8a525f165c2b36072bc5b11096f0d1fc6eaf103d). It seems to be te latest. I also ran `:PlugUpgrade` and it says "vim-plug is already up-to-date".

**Comment by andreicristianpetcu** [comment]
I think it's an old vim, I think my box uses some special kind of repositoryes and pulls an old Vim even if ti is a fairly recent CentOS. What do you think?

**Comment by junegunn** [comment]
Okay, thanks.

@starcraftman Can you look into this? This seems related to the python installer.

```
+python +quickfix +reltime +rightleft -ruby
```

**Comment by junegunn** [comment]
@andreicristianpetcu Meanwhile you can try using the previous revision w/o python installer, or use the single-threaded Vim installer with `:PlugInstall1`

**Comment by andreicristianpetcu** [comment]
Than you @junegunn I used this revision 74169f37610aedef7d0ec1e2b26948d474d3f9a0 and it works only not in paralel. Cool stuff! If I do not have python or ruby does it install the plugins or are python and ruby hard dependencies?

**Comment by vheon** [comment]
@andreicristianpetcu only the parallel install/update have a hard dependency on ruby/python

**Comment by junegunn** [comment]
They are optional. And you don't need both at the same time. 
1. If your Vim is compiled with ruby support, vim-plug will use it to install plugins,
2. if it's not, vim-plug will check if python support is available and use that version of the installer when available.
3. If none of them is supported, plugins will be installed using single-threaded vimscript installer.

**Comment by vheon** [comment]
@andreicristianpetcu if you put `let g:plug_log_on = 1` in your vimrc then there will be a `plug_logs` directory in your home with all the logging. If you can post them will help @starcraftman debug this :+1:

**Comment by andreicristianpetcu** [comment]
Thank you for your explinations! @vheon @junegunn 

@starcraftman I put this in my ~/.vimrc but `ls -la ~/`does not show any files related to plug. I see .vim and .vimrc but no plug_logs.

```vimscript
let mapleader = "\<Space>"                                                                           
let g:plug_log_on = 1                                                                                
if has('vim_starting')
```
https://github.com/andreicristianpetcu/dotfiles/blob/c6ed5186f8d0e1d0e31cded252185d40bb4fdb94/vimrc#L3

**Comment by starcraftman** [comment]
@andreicristianpetcu Just like the recent issues ruby ran into, when nothing displays it often indicates the interpreter failed early on before any buffer updates. Logs require the interpreter actually get to that step, seems it is failing on something basic like imports or some call not available. I can't guarantee this will be fixed quickly.

For now, to prevent python working, open the plug.vim file & set `let s:py2 = 0`.

@junegunn You may be right, may have to do version checking like ruby does. Was inevitable I suppose.

**Comment by andreicristianpetcu** [comment]
@starcraftman right now I use older code that works single threaded, please tell me if you need any more info from me. From what I understand the `let s:py2 = 0` suggestion was a workaround, not a troubleshooting step.

How can I provide logs for imports or calls that are not available? Thank you!

**Comment by starcraftman** [comment]
@andreicristianpetcu I'll try simple debugging by building 7.2.411 on one of my systems and seeing if it works with just +python. Similarly trying python 2.6.6, I'm using some 2.7.x version. It is probably one or the other. I'll ping you if any extra information required. Thanks for the offer. :)

**Comment by vheon** [comment]
@starcraftman @andreicristianpetcu there is no need to modify the `plug.vim` to force the single threaded version, just put `let g:plug_threads = 1` in your vimrc and you'l be using the threaded version written in viml.

**Comment by starcraftman** [comment]
@vheon Right you are, I forgot about that check. Better solution.

**Comment by starcraftman** [comment]
Note while I am investigating, this is what is currently proven to work by tests.

Vim Version | Python Version |  Context
-------------- | -------------------- | -------------
| Latest | 2.7.3 | Travis Test |
| Latest | 2.7.6 | My Setup |

**Comment by junegunn** [comment]
@andreicristianpetcu The vagrant box you linked has Vim 7.4.160 installed with `+python/dyn` and `+ruby/dyn`, and I couldn't reproduce the problem. Did you install Vim 7.2 on purpose?

```ruby
  config.vm.box = "centos7"
  config.vm.box_url = "https://f0fff3908f081cb6461b407be80daf97f07ac418.googledrive.com/host/0BwtuV7VyVTSkUG1PM3pCeDJ4dVE/centos7.box"
```

**Comment by starcraftman** [comment]
@junegunn He mentioned above. I assume there is some reason his project/team doesn't simply use more up to date.

> I think it's an old vim, I think my box uses some special kind of repositoryes and pulls an old Vim even if ti is a fairly recent CentOS. What do you think?

**Comment by junegunn** [comment]
@starcraftman @andreicristianpetcu Ah thanks, I missed it.

But anyway I can reproduce the same on Ubuntu 10.04 vagrant box. v:exception say `Vim(python):  File "<string>", line 109`.

```ruby
  config.vm.box = "lucid"
  config.vm.box_url = "https://dl.dropbox.com/u/14741389/vagrantboxes/lucid64-lamp.box"
```

```
:version
VIM - Vi IMproved 7.2 (2008 Aug 9, compiled Nov 15 2012 00:34:01)
Included patches: 1-330
Compiled by buildd@
Huge version without GUI.  Features included (+) or not (-):
+arabic +autocmd -balloon_eval -browse ++builtin_terms +byte_offset +cindent -clientserver -clipboard +cmdline_compl +cmdline_hist +cmdline_info +comments +cryptv +cscope
+cursorshape +dialog_con +diff +digraphs -dnd -ebcdic +emacs_tags +eval +ex_extra +extra_search +farsi +file_in_path +find_in_path +float +folding -footer +fork() +gettext
-hangul_input +iconv +insert_expand +jumplist +keymap +langmap +libcall +linebreak +lispindent +listcmds +localmap +menu +mksession +modify_fname +mouse -mouseshape +mouse_dec
+mouse_gpm -mouse_jsbterm +mouse_netterm -mouse_sysmouse +mouse_xterm +multi_byte +multi_lang -mzscheme -netbeans_intg -osfiletype +path_extra +perl +postscript +printer
+profile +python +quickfix +reltime +rightleft +ruby +scrollbind +signs +smartindent -sniff +startuptime +statusline -sun_workshop +syntax +tag_binary +tag_old_static
-tag_any_white +tcl +terminfo +termresponse +textobjects +title -toolbar +user_commands +vertsplit +virtualedit +visual +visualextra +viminfo +vreplace +wildignore +wildmenu
+windows +writebackup -X11 -xfontset -xim -xsmp -xterm_clipboard -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
      user exrc file: "$HOME/.exrc"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -Wall -g -O2    -D_REENTRANT -D_GNU_SOURCE -DDEBIAN  -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET
_BITS=64  -I/usr/lib/perl/5.10/CORE  -I/usr/include/python2.6 -pthread -I/usr/include/tcl8.4  -D_REENTRANT=1  -D_THREAD_SAFE=1  -D_LARGEFILE64_SOURCE=1  -I/usr/lib/ruby/1.8/x86_6
4-linux
Linking: gcc   -L. -Wl,-Bsymbolic-functions -rdynamic -Wl,-export-dynamic  -Wl,-E  -Wl,--as-needed -o vim       -lncurses -lselinux  -lacl -lgpm -Wl,-E  -fstack-protector -L/usr/
local/lib  -L/usr/lib/perl/5.10/CORE -lperl -L/usr/lib/python2.6/config -lpython2.6 -lutil -Xlinker -export-dynamic -Wl,-O1 -Wl,-Bsymbolic-functions -L/usr/lib -ltcl8.4 -lieee -l
ruby1.8 -lrt -lm
```

**Comment by junegunn** [comment]
1. https://github.com/junegunn/vim-plug/blob/master/plug.vim#L1102 is Python 2.7+ syntax
2. I removed the line and enabled `g:plug_log_on`, but ~/plug_logs/main is empty.
3. The code fails in `GLog.write` call. This seems to be the problem

```python
:python print '{}'.format("foo")
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ValueError: zero length field name in format
```

**Comment by starcraftman** [comment]
@junegunn I see, disappointing, I guess something to do with my using inline set instead of list. I shall try booting up this lucid image and debugging everything on that. Probably easiest.

Regarding GLog, I imagine similarly that the use of {} in format calls is a 2.7 thing. One of my favorite features! :(

**Comment by junegunn** [comment]
@starcraftman I see, it's your decision to support 2.6 or not. But in either way, I think it would be nice if we need check the version of python. (The same can be told about Ruby, but I can't imagine anyone using Ruby < 1.8)

```patch
diff --git a/plug.vim b/plug.vim
index a6a4e63..f2b5f44 100644
--- a/plug.vim
+++ b/plug.vim
@@ -236,18 +236,22 @@ function! s:trim(str)
   return substitute(a:str, '[\/]\+$', '', '')
 endfunction
 
-function! s:git_version_requirement(...)
-  let s:git_version = get(s:, 'git_version',
-    \ map(split(split(s:system('git --version'))[-1], '\.'), 'str2nr(v:val)'))
-  for idx in range(0, a:0 - 1)
-    let v = get(s:git_version, idx, 0)
-    if     v < a:000[idx] | return 0
-    elseif v > a:000[idx] | return 1
+function! s:version_requirement(val, min)
+  for idx in range(0, len(a:min) - 1)
+    let v = get(a:val, idx, 0)
+    if     v < a:min[idx] | return 0
+    elseif v > a:min[idx] | return 1
     endif
   endfor
   return 1
 endfunction
 
+function! s:git_version_requirement(...)
+  let s:git_version = get(s:, 'git_version',
+    \ map(split(split(s:system('git --version'))[-1], '\.'), 'str2nr(v:val)'))
+  return s:version_requirement(s:git_version, a:000)
+endfunction
+
 function! s:progress_opt(base)
   return a:base && !s:is_win &&
         \ s:git_version_requirement(1, 7, 1) ? '--progress' : ''
@@ -755,6 +759,14 @@ function! s:update_impl(pull, force, args) abort
   call append(0, ['', ''])
   normal! 2G
 
+  " Python version requirement (>= 2.7)
+  if s:py2 && !s:ruby && !s:nvim && s:update.threads > 1
+    redir => pyv
+    silent python import platform; print(platform.python_version())
+    redir END
+    let s:py2 = s:version_requirement(
+          \ map(split(split(pyv)[0], '\.'), 'str2nr(v:val)'), [2, 7])
+  endif
   if (s:py2 || s:ruby) && !s:nvim && s:update.threads > 1
     try
       let imd = &imd
```

**Comment by andreicristianpetcu** [comment]
@junegunn My setup is not that trivial. It is a Centos connecting to a puppet master and pulls lots of complicated stuff. That's how I end up with an old vim. I'm glad you can reproduce it on an old Ubuntu. I think it can be reproduced on Centos6 also.

I use [pyenv](https://github.com/yyuu/pyenv) to run different python versions when this is needed. It is really simple to use (a lot better than virtualenv)

**Comment by vheon** [comment]
@starcraftman is not the `{}` per se, but in python 2.6 you have to use indexes so `"{0}".format` is not the prettiest thing but is not that bad.

**Comment by starcraftman** [comment]
@junegunn I haven't decided specifically whether we'll support 2.6, depends how much is broken/missing. I meant by debug more that I'd explore the amount of breakage to fix on this lucid 2.6 image.

Re version, it was inevitable. Something like that will probably be used. Are you in a rush to prevent any issues? If so you can put that in, don't forget the `!s:is_win` to prevent windows users.

@vheon Yes, that's what I meant, I guess it was unclear as I wrote. 2.7 introduced the '{}' formatting that makes formatting life much easier.

**Comment by vheon** [comment]
@junegunn oh ok :+1:

**Comment by junegunn** [comment]
@starcraftman The initial definition of `s:py2` already checks `!s:is_win`. The code above reassigns the value just before starting python installer (I didn't want to execute python code until it's used). I'm going to put the code and close the issue for now. If you decide that you're going to support python 2.6, please send me another PR.

**Comment by starcraftman** [comment]
@junegunn Fine by me for now, will prevent issues while investigating.

**Comment by junegunn** [comment]
@andreicristianpetcu I disabled python installer on python 2.6 or below, so after a `:PlugUpgrade`, it should automatically choose non-parallel installer on that environment (no need for `g:plug_threads = 1`)

> I use pyenv to run different python versions when this is needed. It is really simple to use (a lot better than virtualenv)

Thanks for the info, pyenv looks interesting. Python is not my language of choice, but I do use virtualenv for running Ansible. I think I should try pyenv.

@starcraftman Thank you!

**Comment by andreicristianpetcu** [comment]
Thank you all for fixing this! I tested on my Centos machine and it fallsback to singlethreaded version when Python is old and in Arch it flies to the multithreaded version! :D :+1:
